8231601: Update CONTRIBUTING.md to clarify process for contributing features plus Skara changes

kevinrushforth opened the pull request:
This PR updates the `CONTRIBUTING.md` guide to address the following:

1. Clarify the process for adding new features / API changes, specifically that they must be discussed on the mailing list as the first step.
2. Add a link to the mailing list in the section regarding contributing bug fixes.
3. Remove the text about cross-linking the PR and JBS issue, and add a note that the Skara tooling takes care of this
4. Remove the section about manually resolving the JBS issue, and add a note that the Skara bot automatically does this when the PR is integrated.
5. Suggest the use of the "/reviewers 2" and "/csr" commands when appropriate
6. Update the note regarding which JDK(s) to use.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8231601](https://bugs.openjdk.java.net/browse/JDK-8231601): Update CONTRIBUTING.md to clarify process for contributing features plus Skara changes


### Reviewers
 * [Nir Lisker](https://openjdk.java.net/census#nlisker) (@nlisker - **Reviewer**)
 * [Johan Vos](https://openjdk.java.net/census#jvos) (@johanvos - **Reviewer**)
 * [Michael Strauß](https://openjdk.java.net/census#mstrauss) (@mstr2 - Author)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/303/head:pull/303` \
`$ git checkout pull/303`

Update a local copy of the PR: \
`$ git checkout pull/303` \
`$ git pull https://git.openjdk.java.net/jfx pull/303/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 303`

View PR using the GUI difftool: \
`$ git pr show -t 303`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/303.diff">https://git.openjdk.java.net/jfx/pull/303.diff</a>

</details>

nlisker reviewed (commented):
Additional comments: 

1. "Use Unix-style (LF) line endings not DOS-style (CRLF)" needs a comma before "not".
2. "Line width is no more than 120 characters" I remember that it was 130 or 135 somewhere.
3. "Wildcard imports (import foo.bar.baz.*) are forbidden" Junit imports use them extensively.
4. `./gradlew all test` will cause failure on webkit tests if it was not built.

kevinrushforth commented on CONTRIBUTING.md:
```diff
@@ -15,13 +15,13 @@ Provide as much information as you can. The easier it is for us to recreate your
 
-If you find yourself wishing for a feature that doesn't exist in OpenJFX, you are probably not alone. There are bound to be others out there with similar needs. Many of the features that OpenJFX has today have been added because our users saw the need.
-Open an issue on our [issues list](https://bugs.openjdk.java.net) on JBS which describes the feature you would like to see, why you need it, and how it should work.
+All new feature requests, including any API changes, need prior discussion on the [openjfx-dev](mailto:openjfx-dev@openjdk.java.net) mailing list, even if there is already an open
```
Fixed.

kevinrushforth commented on CONTRIBUTING.md:
```diff
@@ -15,13 +15,13 @@ Provide as much information as you can. The easier it is for us to recreate your
-If you find yourself wishing for a feature that doesn't exist in OpenJFX, you are probably not alone. There are bound to be others out there with similar needs. Many of the features that OpenJFX has today have been added because our users saw the need.
-Open an issue on our [issues list](https://bugs.openjdk.java.net) on JBS which describes the feature you would like to see, why you need it, and how it should work.
+All new feature requests, including any API changes, need prior discussion on the [openjfx-dev](mailto:openjfx-dev@openjdk.java.net) mailing list, even if there is already an open
+[JBS issue](https://bugs.openjdk.java.net). See the [New features / API additions](#new-features--api-additions) section at the end of this guide for more information.
```
Fixed (I included a link to a filter the first time, where it suggests to search for already-reported issues).

kevinrushforth commented on CONTRIBUTING.md:
```diff
@@ -15,13 +15,13 @@ Provide as much information as you can. The easier it is for us to recreate your
 -------------------------------------------
 
-If you have a bug fix or new feature that you would like to contribute to OpenJFX, please find or open an issue about it first. Talk about what you would like to do. It may be that somebody is already working on it, or that there are particular issues that you should know about before implementing the change. Feature requests, in particular, should be discussed ahead of time and will require sign ...
+If you have a bug fix or new feature that you would like to contribute to OpenJFX, please find or open an issue about it first. Talk about what you would like to do on the [openjfx-dev](mailto:openjfx-dev@openjdk.java.net) mailing list. It may be that somebody is already working on it, or that there are particular issues that you should know about before implementing the change. Feature requests, ...
```
I think I'll remove the comment about "please find or open an issue about it first" and just say:

```
If you have a bug fix or new feature that you would like to contribute to OpenJFX, please talk about what you would like to do...
```

kevinrushforth commented on CONTRIBUTING.md:
```diff
@@ -94,12 +95,6 @@ Once your changes and tests are ready to submit for review:
-
     Please adhere to the general guideline that you should never force push
     to a publicly shared branch. Once you have opened your pull request, you
     should consider your branch publicly shared. Instead of force pushing
```
I added a pointer to the GitHub help page here.

kevinrushforth commented on CONTRIBUTING.md:
```diff
@@ -94,12 +95,6 @@ Once your changes and tests are ready to submit for review:
     ready for it to be reviewed. The Skara bot will not send an RFR
     email unless the title starts with a 7-digit bug ID.
```
I updated the doc to mention Draft PRs

kevinrushforth commented:
Regarding your additional comments:

> * "Use Unix-style (LF) line endings not DOS-style (CRLF)" needs a comma before "not".

Fixed.

> * "Line width is no more than 120 characters" I remember that it was 130 or 135 somewhere.

You're probably remembering an old version, but it's been 120 for a while now.

> * "Wildcard imports (import foo.bar.baz.*) are forbidden" Junit imports use them extensively.

Fixed to add an exception for wildcard static imports in tests.

> * `./gradlew all test` will cause failure on webkit tests if it was not built.

Added a note about this and a pointer to the [Web Testing](WEBKIT-MEDIA-STUBS.md) doc.

kevinrushforth commented:
/reviewers 2

nlisker commented:
The "New features / API additions" repeats some things already stated. Is it to make each section independent?

kevinrushforth commented:
I wanted the "New features / API additions" section to stand on its own. Once thing that might be redundant now is the following sentence in the "Contributing code and documentation changes" section:

"Feature requests, in particular, must be discussed ahead of time and will require significant effort on your part."

I think that could be removed or incorporated in "New features / API additions".

nlisker commented:
I think that "Contributing to the OpenJFX codebase" should be renamed to something related to a style guide. Then split the testing part to its own subsection.

Also, I still find it confusing that "New features / API additions" is not under the code contribution section. There seems to be 2 main sections: reporting bugs / requesting features - these don't involve code, just talk; then there is contributing code, which covers the process for setup, submissions of bugs fixes, submission of features/API, style, and testing (in some order). Wouldn't this be a better flow?

kevinrushforth commented:
Yes, I do think the flow could be better. I'll need to put this on hold for a while, but when I get back to it, I'll look at your suggestions and see if I can come up with something that will improve the flow.

Btw, the thinking behind putting the "New features / API additions" sections at the end (sort of like an appendix) is that I didn't want it to get in the way of the "here's how you sumbit and review a PR" for bug fixes, which is the more common case. I don't think it achieves that in its current form.

lorenzpahl commented on CONTRIBUTING.md:
```diff
@@ -207,8 +193,9 @@ Please also follow these formatting guidelines:
-* Wildcard imports (`import foo.bar.baz.*`) are forbidden and may cause the build to fail. Please attempt to tame your IDE so it doesn't make them and please send a PR against this document with instructions for your IDE if it doesn't contain them.
+* Avoid making changes that are unrelated to the bug you are fixing. This includes fixing minor errors such as warnings, spacing / indentation, spelling errors, etc, in code that you aren't otherwise modifying as part of your fix.
+* Disable &ldquo;auto-format on save&rdquo; to prevent your IDE from making unnecessary formatting changes. This makes reviews much harder as it generates unnecessary diffs. If your IDE supports formatting only modified chunks, that is fine to do.
+* Wildcard imports &ndash; for example, `import java.util.*;` &ndash; are forbidden and may cause the build to fail. Please attempt to configure your IDE so it doesn't genearte wildcard imports. An exception to this rule is that wildcard static imports in test classes are allowed, for example, `import static org.junit.Assert.*;`.
```
There is a small spelling error:  "genearte".

kleopatra commented:
not sure whether it belongs here, or whether or not it's obviously implied but: I would like to see a bit of clarification on testing of contributions. Right now the sentence might be interpreted to be about running available tests:

> Test your changes
> Run the test suite to make sure that nothing is broken.
> 

add something like:

     For most code changes, new tests covering those changes are mandatory.

Maran23 commented:
Is there any plan to integrate this? I quite like the change and it makes the process much more clear. :)

kevinrushforth commented:
Yes, I'll pick this back up soon.

kevinrushforth commented:
This has been dormant for a long time, so I just merged the latest master. I also discovered a couple additional changes I had in a "working" branch that I will push to address a couple minor comments.

I still need to get back to this to address the comments from @nlisker about the flow, so that might take a bit more time, but I would like to get this moving to at least fix the fact that several things are wrong or misleading in the current doc.

johanvos commented on CONTRIBUTING.md:
```diff
@@ -161,43 +165,25 @@ Once your changes and tests are ready to submit for review:
-your JDK 11 installation. By default, tests use the same runtime as `JAVA_HOME`.
-Currently OpenJFX builds are running on JDK 11 (recommended) and JDK 12.
+your JDK installation. By default, tests use the same runtime as `JAVA_HOME`.
+Currently OpenJFX builds are running on JDK 11 through JDK 15. JDK 14 is recommended.
```
JDK 17

kevinrushforth replied:
Done.

johanvos commented on CONTRIBUTING.md:
```diff
@@ -54,15 +54,15 @@ If you are a first time contributor to OpenJFX, welcome! Please do the following
     A developer with an active OpenJDK ID can file a bug directly in JBS.
 
-TIP: A GitHub pull request (PR) should not be the first time we hear about your proposed change to OpenJFX. Unless clearly identified as experimental or work-in-progress (WIP), we will usually close a pull request that isn't associated with an existing bug report. Reading the policies below will help you in getting your change approved.
+TIP: A GitHub pull request (PR) should not be the first time we hear about your proposed change to OpenJFX. Unless clearly identified as a Draft or work-in-progress (WIP), we will usually close a pull request that isn't associated with an existing bug report. Reading the policies below will help you in getting your change approved.
```
forward reference to Draft or WIP, so maybe add (as explained later)?

kevinrushforth replied:
Done.

kevinrushforth commented on CONTRIBUTING.md:
```diff
@@ -223,5 +210,43 @@ If you are changing anything that might possibly affect rendering, you should ru
+that the final review / approval of the CSR usually happens late in the review cycle.
+
+TIP: a pull request is _not_ the starting point, since that skips the first two important steps and jumps right into
+"given this new feature, and an API definition that specifies it, please review my proposed implementation".
```
This is a good idea, and is part of what we mean when asking contributors of features to think about "stewardship". So I think adding it either here or in the code review section of the Wiki is good.

kevinrushforth commented on CONTRIBUTING.md:
```diff
@@ -161,43 +165,53 @@ Once your changes and tests are ready to submit for review:
+----------------------------
+
+Adding a new feature to OpenJFX requires us to consider what it means to support that API forever; we take compatibility seriously.
+The main idea is to think in terms of "stewardship" when evolving the JavaFX API. This will require significant effort on your part.
```
@johanvos to address your point about maintenance, what do you think about adding a sentence here? Something like:

This will require significant effort on your part. __It begins before you submit a pull request for review, and continues after the new feature is integrated.__

We can follow this up with more detail in the code review guidelines.

johanvos replied:
That's a good sentence. It clearly and concisely tells what it is expected.

johanvos reviewed (approved):
This PR is a major improvement over the existing status. 
I therefore recommend to integrate this, and have other issues being discussed in follow-up issues. A contributing guideline document is a moving target.

kevinrushforth commented:
> I therefore recommend to integrate this, and have other issues being discussed in follow-up issues

Agreed. I think I've addressed all comments made in this PR.

@aghaisas @arapte @nlisker can one of you be the second reviewer?

nlisker commented:
I will review this shortly.

mstr2 commented on CONTRIBUTING.md:
```diff
@@ -15,13 +15,13 @@ Provide as much information as you can. The easier it is for us to recreate your
 Feature requests
 ----------------
 
-If you find yourself wishing for a feature that doesn't exist in OpenJFX, you are probably not alone. There are bound to be others out there with similar needs. Many of the features that OpenJFX has today have been added because our users saw the need.
```
I think you should restore this sentence. It feels uplifting and encouraging, while the other added sections feel pretty discouraging at first sight for newcomers to the project.

nlisker replied:
Fair point. I don't mind restoring it.

kevinrushforth replied:
OK, I can add this back, although I might reword it slightly.

nlisker commented on CONTRIBUTING.md:
```diff
@@ -161,43 +165,54 @@ Once your changes and tests are ready to submit for review:
+3. Submit a review of your proposed implementation. As noted in the
+[New features / API additions](https://wiki.openjdk.java.net/display/OpenJFX/Code+Reviews#CodeReviews-NewFeaturesC.Newfeatures/APIadditions.)
+section of the Code Review Policies doc, we also need a [CSR](https://wiki.openjdk.java.net/display/csr/Main), which documents the API change and its approval.
+The CSR can be reviewed in parallel. Changes in the API that arise during the review need to be reflected in the CSR, meaning
+that the final review / approval of the CSR usually happens late in the review cycle.
```
I would advise contributors to not deal with the CSR until the review is almost done. It just duplicates the places that you need to update. I'm saying this from personal experience :)

kevinrushforth replied:
Good idea.

nlisker reviewed (approved):
I added 1 suggestion.

kevinrushforth commented:
/integrate
